RipCurrent/ripcurrent.h

```diff
@@ -0,0 +1,659 @@
+      : T_BASE(spawner->GetUniqueDefinition()), shared_impl_spawner_(spawner) {}
+
+  std::shared_ptr<InstanceBeingRun<INPUT, OUTPUT>> SpawnAndRun(
+      std::shared_ptr<EntriesConsumer<NEXT_INPUT>> next) const override {
+    return shared_impl_spawner_->SpawnAndRun(next);
+  }
+
+  // User-facing `RipCurrent()` method.
+  template <InputPolicy IN = INPUT, OutputPolicy OUT = OUTPUT>
+  typename std::enable_if<IN == InputPolicy::DoesNotAccept && OUT == OutputPolicy::DoesNotEmit,
+                          RipCurrentRunContext>::type
+  RipCurrent() {
+    SpawnAndRun(std::make_shared<DummyNoEntryTypeAcceptor>());
+
+    // TODO(dkorolev): Return proper run context. So far, just require `.Sync()` to be called on it.
+    std::ostringstream os;
+    os << "RipCurrent run context was left hanging. Call `.Sync()`, or, well, something else. -- D.K.\n";
+    T_BASE::GetDefinition().FullDescription(os);
+    return std::move(RipCurrentRunContext(os.str()));
+  }
+
+ private:
+  std::shared_ptr<T_BASE> shared_impl_spawner_;
+};
+
+// Helper code to initialize the next handler in the chain before the user code is constructed.
+// TL;DR: The user should be able to use `emit()` from constructor, no strings attached. Thus,
+// the destination fo this `emit()` should be initialized before the user code is.
+template <InputPolicy INPUT>
+class NextHandlerContainer {
+ public:
+  virtual ~NextHandlerContainer() = default;
+
+  void SetNextHandler(std::shared_ptr<EntriesConsumer<INPUT>> next) const { next_handler_ = next.get(); }
+
+ protected:
+  void emit(int x) const { next_handler_->Accept(x); }
+
+ private:
+  // NOTE(dkorolev): This field is pre-initialized by an instance of another helper class, which is
+  // constructed before the constuctor of the user code, derived indirectly from `NextHandlerContainer`,
+  // is run. This is done to enable user code to do `emit()` from within the constructor, or from callbacks
+  // initialized from within the constructor (for example, HTTP endpoints).
+  // Since such an initialization happens before the constructor has been properly called,
+  // the data field it has to be a POD. An `std::shared_ptr<>` would not work due to being uninitialized first
+  // (and re-initialized over later).
+  // Thus, we use a bare pointer here. Note that the `Instance` classes, that ultimately instantiate user code,
+  // do wrap the top-level `next` into its own `std::shared_ptr<>` and make sure to construct it before (and
+  // destruct it after) the user code itself.
+  mutable EntriesConsumer<INPUT>* next_handler_;
+};
+
+template <InputPolicy INPUT, typename T>
+class NextHandlerInitializer {
+ public:
+  template <typename... ARGS>
+  NextHandlerInitializer(std::shared_ptr<EntriesConsumer<INPUT>> next, ARGS&&... args)
+      : early_initializer_(impl_, next), impl_(std::forward<ARGS>(args)...) {}
+
+  void Accept(const int& x) { impl_.f(x); }  // TODO(dkorolev): This won't be `int` very soon.
```

Comment:
Right now it's `emit(...)` vs. `f(...)`. I thought `f()` is the shortest good name for a in-class member (`operator()` is way too complex), and, since `f` would be lowercase, `emit()` is also best lowercase.

Open for suggestions. `Emit(...)` LGTM right away, re. use code defining `Accept()`, I'm a bit uneasy about its length.